indimail-x/vgroup.1
.TH vgroup 1
.SH NAME
vgroup \- create distribution groups for set of users in a virtual domain

.SH SYNOPSYS
.nf
\fBvgroup\fR \fB\-a\fR [\fB\-n\fR] [\fB\-c\fR][\fBq\fR] [\fB\-v\fR] \fIgroupAddress\fR [\fIpassword\fR]
\fBvgroup\fR [\fB\-i\fR \fIaliasLine\fR] [\fB\-d\fR \fIaliasLine\fR]
  [\fB\-u\fR \fInewaliasLine\fR \fB\-o\fR \fIoldaliasLine\fR] [\fB\-v\fR] \fIgroupAddress\fR
.fi

.SH DESCRIPTION
\fBvgroup\fR implements groups functionality in IndiMail. It is a utility to add groups.
On an existing group, vgroup can be used to add, modify and delete members.

The first form of usage is used for creating a new group. Subsequently normal IndiMail programs
(\fBvuserinfo\fR, \fBvdeluser\fR, \fBvmoduser\fR, \fBvpasswd\fR, \fBvsetuserquota\fR) can be
used to operate on a group exactly like on virtual users. In fact, a group in IndiMail is
a virtual user.

The second form of usage is used to add, modify, delete aliases to the group. It uses the
\fBvalias\fR to do this. The most comman usage of vgroup is to add forwardings. 
i.e. All mails sent to the group will be forwarded to members. In this case address will
be of the form '&email_address'.

The implementation of groups in IndiMail is actually a normal email account with aliases.
Hence a group can also be created using \fBvadduser\fR in conjunction with \fBvalias\fR with
the restriction that the gecos starts with the word 'MailGroup' (see option \fB\-c\fR below for
more explanation). vgroup has been written for user convenience alone.

.SH OPTIONS
.PP
Common Options
.TP 3
\fB\-v\fR
Sets verbose mode while vgroup is operating

.TP
\fIgroupAddress\fR
Email address for the group.

.PP
Options for adding a New Group.
.TP 3
\fB\-e\fR
Set the encrypted Password field
This options disables the internal encryption routine which encrypts the
password provided on the command line. This option sets the encrypted
password field exactly as given on the command line without any encryption.
It expects you to give a standard encrypted password or you can use
this to set plaintext/salted password for CRAM authentication.

.TP
\fB\-r\fR \fIlen\fR
Generates Random Password of length \fIlen\fR. Need not give password on
command line.

.TP
\fB\-h\fR \fIhash\fR
Specify hash which is one of DES, MD5, SHA-256, SHA-512. Here \fIID\fR is
the id in \fI/etc/shadow\fR. See \fBincrypt\fR(1).
.TS
l l.
HASH    ID Description
_
DES     -  DES encryption (shouldn't be used)
MD5     1  MD5 encryption (shouldn't be used)
SHA-256 5  SHA256 encryption
SHA-512 6  SHA512 encryption
.TE

You can also set the environment variable \fBPASSWORD_HASH\fR to set the
encryption method. The -h argument overrides the environment variable
\fBPASSWORD_HASH\fR. The value of \fBPASSWORD_HASH\fR environment variable
identifies the encryption method used and this then determines how the rest
of the password string is interpreted. The following values of
\fBPASSWORD_HASH\fR are supported:

.nf
.ta 5c 10c
PASSWORD_HASH Value	Description
0	DES encryption (shouldn't be used)
1	MD5 encryption (shouldn't be used)
2	SHA256 encryption
3	SHA512 encryption
.fi
.RE

.TP 3
\fB\-m\fR \fIscram\fR
Sets the CRAM or SCRAM method for encryption. This will set SCRAM password in
the \fIscram\fR field in indimail/indibak tables. For CRAM method, it will
set clear text password when -C option is specified.
.RS
.nf
.ta 5c 10c
CRAM/SCRAM method	Description
-----------------	-----------
CRAM	Sets clear text password suitable for any
	CRAM method (CRAM-MD5, CRAM-SHA1, ...)
SCRAM-SHA-1	SHA1 encryption suitable for SCRAM-SHA-1.
SCRAM-SHA-256	SHA256 encryption suitable for SCRAM-SHA-256.
.fi
.tc
.RE

.TP
\fB\-C\fR
Sets up authentication suitable for CRAM-MD5, CRAM-SHA1, CRAM-SHA224,
CRAM-SHA256, CRAM-SHA384, CRAM-SHA512, CRAM-RIPEMD and DIGEST-MD5 methods.
This works by storing the clear text credentials in the database. if the
\-\fIm\fR option is selected, this will additionally store a hex-encoded
salted password for SCRAM methods, which can be used instead of clear text
passwords by clients (for SCRAM authentication).

.TP 3
\fB\-S\fR \fIsalt\fR
Specify a base64 encoded salt to be used when generating SCRAM password. If
not specified, this will be generated using libsodium/gsasl. Here base64
implies characters [0-9], [a-z], [A-Z] and the two characters [./].

.TP
\fB\-I\fR \fIiteration\fR
Specify the iteration count to be used when generating SCRAM password. The
default is 4096.

.TP
\fIpasswd\fR
Passwd for the user. If not given, user will be prompted.

.TP
\fB\-a\fR
Add a new group.

.TP
\fB\-H\fR \fIhostid\fR
For a clustered domain, this option can be used to create the user on a specific host having
\fIhostid\fR as the hostid.

.TP
\fB\-M\fR \fImdahost\fR
For a clustered domain, this option can be used to create the user on a specific cluster
having Mail Delivery Host as \fImdahost\fR.

.TP
\fB\-c\fR
Comment field (gecos) for the group account. Internally the comment starts with the
word 'IndiGroup'. This will be shown if you use the command vuserinfo on the group address.
It is advisable not to change this by external programs such as vmoduser. IndiMail figures
out difference between a normal email account and a group account by checking if gecos
field starts with the word 'IndiGroup'.

.TP
\fB\-q\fR
Sets the Quota for the group. This should be set in case the group will also function as a
normal user and receive mails.

.TP
\fB\-n\fR
Ignore requirement of groupAddress to be a local address on the node where vgroup is
being run (option valid for clustered domain only)

.PP
Options for an Existing Group
.TP
\fB\-i\fR \fIaliasLine\fR
Insert alias line \fIaliasLine\fR in the group. An alias line has one of the 3 forms.
See \fBdot-qmail(5)\fR for more details.

.EX
 1. &\fIemail\fR       - Forward mails to \fIemail\fR
 2. Maildir      - copy mails to a Maildir with path \fIMaildir\fR
 3. | Command    - pipe the mail through command \fICommand\fR
.EE

.TP
\fB\-d\fR \fIaliasLine\fR
delete aliasLine from the group. See valias for more details

.TP
\fB\-u\fR \fInewaliasLine\fR \fB\-o\fR \fIoldaliasLine\fR
Modify oldaliasLine to newaliasLine alias in the group. See valias for more details

.SH RETURN VALUE
0 if all steps were successful, non-zero otherwise. If any of the steps fail, a
diagnostic message is printed.

.SH "SEE ALSO"
dot-qmail(5), valias(1), vcalias(8), vadduser(1), vdelivermail(8), vrenamedomain(1), vreorg(8)


indimail-x/NEWS
doc/RELEASE-Notes-indimail.md

indimail-x/vgroup.c
/*
 * $Id: vgroup.c,v 1.8 2023-07-17 12:28:44+05:30 Cprogrammer Exp mbhangui $
 */
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#ifndef	lint
static char     sccsid[] = "$Id: vgroup.c,v 1.8 2023-07-17 12:28:44+05:30 Cprogrammer Exp mbhangui $";
#endif

#ifdef VALIAS
#define XOPEN_SOURCE = 600
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#ifdef HAVE_QMAIL
#include <sgetopt.h>
#include <stralloc.h>
#include <strerr.h>
#include <str.h>
#include <fmt.h>
#include <scan.h>
#include <env.h>
#include <makesalt.h>
#include <hashmethods.h>
#include <subfd.h>
#endif
#ifdef HAVE_GSASL_H
#include <gsasl.h>
#include "gsasl_mkpasswd.h"
#endif
#include "parse_email.h"
#include "sqlOpen_user.h"
#include "iopen.h"
#include "iclose.h"
#include "iadduser.h"
#include "sql_getpw.h"
#include "get_real_domain.h"
#include "parse_quota.h"
#include "sql_getip.h"
#include "SqlServer.h"
#include "vgetpasswd.h"
#include "strmsg.h"
#include "valias_insert.h"
#include "valias_delete.h"
#include "valias_update.h"
#include "common.h"
#include "variables.h"

#define ADDNEW_GROUP  0
#define INSERT_MEMBER 1
#define DELETE_MEMBER 2
#define UPDATE_MEMBER 3
#define FATAL         "vgroup: fatal: "
#define WARN          "vgroup: warning: "

static char    *usage =
	"usage1: vgroup -a [-cqvVhmI] groupAddress [password]\n"
	"usage2: vgroup    [-iduovV]  groupAddress\n\n"
	"options\n"
	"  -V                - print version number\n"
	"  -v                - verbose\n"
	"  -a                - add new group\n"
	"  -r len            - generate a random password of length=len\n"
	"  -e password       - set the encrypted password field\n"
	"  -h hash           - use one of DES, MD5, SHA256, SHA512, YESCRYPT, hash method\n"
#ifdef HAVE_GSASL
#if GSASL_VERSION_MAJOR == 1 && GSASL_VERSION_MINOR > 8 || GSASL_VERSION_MAJOR > 1
	"  -m SCRAM method   - use one of SCRAM-SHA-1, SCRAM-SHA-256 SCRAM method\n"
	"  -C                - store clear txt and SCRAM hex salted password in database\n"
	"                      This allows CRAM methods to be used\n"
	"  -S salt           - use a fixed base64 encoded salt for generating SCRAM password\n"
	"                    - if salt is not specified, it will be generated\n"
	"  -I iter_count     - use iter_count instead of 4096 for generating SCRAM password\n"
#else
	"  -C                - store clear txt password in database\n"
#endif
#else
	"  -C                - store clear txt password in database\n"
#endif
#ifdef CLUSTERED_SITE
	"  -H hostid         - host on which the group needs to be created - specify hostid\n"
	"  -M mdahost        - host on which the group needs to be created - specify mdahost\n"
	"  -n                - Ignore requirement of of groupAddress to be local\n"
#endif
	"  -c comment        - sets the gecos comment field\n"
	"  -q quota_in_bytes - sets the users quota\n"
	"  -i member_address - insert member to group\n"
	"  -d member_address - delete member from group\n"
	"  -u new -o old     - update old member with a new address)"
	;

static int
get_options(int argc, char **argv, int *option, char **group, char **gecos,
	char **member, char **old_member, char **passwd,
	char **hostid, char **mdahost, char **quota, int *ignore,
	int *encrypt_flag, int *random, int *docram, int *scram, int *iter,
	char **salt)
{
	int             c, i, r;
	char            optstr[27], strnum[FMT_ULONG];

	*group = *gecos = *member = *old_member = *passwd = *hostid =
		*mdahost = *quota = 0;
	*option = -1;
	*ignore = 0;
	*random = 0;
	*encrypt_flag = -1;
	if (docram)
		*docram = 0;
	if (scram)
		*scram = -1;
	if (salt)
		*salt = 0;
	if (iter)
		*iter = 4096;
	/*- make sure optstr has enough size to hold all options + 1 */
	i = 0;
	i += fmt_strn(optstr + i, "vanc:i:d:o:u:q:C", 16);
#ifdef CLUSTERED_SITE
	i += fmt_strn(optstr + i, "H:M:", 4);
#endif
#ifdef HAVE_GSASL
#if GSASL_VERSION_MAJOR == 1 && GSASL_VERSION_MINOR > 8 || GSASL_VERSION_MAJOR > 1
	i += fmt_strn(optstr + i, "m:S:I:", 6);
#endif
#endif
	if ((i + 1) > sizeof(optstr))
		strerr_die2x(100, FATAL, "allocated space for getopt string not enough");
	optstr[i] = 0;

	while ((c = getopt(argc, argv, optstr)) != opteof) {
		switch (c)
		{
		case 'v':
			verbose = 1;
			break;
		case 'r':
			scan_uint(optarg, (unsigned int *) random);
			break;
		case 'a':
			switch (*option)
			{
			case ADDNEW_GROUP:
			case INSERT_MEMBER:
			case DELETE_MEMBER:
			case UPDATE_MEMBER:
				strerr_warn2(WARN, usage, 0);
				return (1);
			}
			*option = ADDNEW_GROUP;
			break;
		case 'n':
			*ignore = 1;
			break;
		case 'c':
			*gecos = optarg;
			break;
		case 'q':
			*quota = optarg;
			break;
		case 'i':
			if (*option == ADDNEW_GROUP) {
				strerr_warn2(WARN, usage, 0);
				return (1);
			}
			if (!*optarg) {
				strerr_warn1("You cannot have an empty email address", 0);
				strerr_warn2(WARN, usage, 0);
				return (1);
			}
			*option = INSERT_MEMBER;
			*member = optarg;
			break;
		case 'd':
			if (*option == ADDNEW_GROUP) {
				strerr_warn2(WARN, usage, 0);
				return (1);
			}
			*option = DELETE_MEMBER;
			*member = optarg;
			break;
		case 'u':
			if (*option == ADDNEW_GROUP) {
				strerr_warn2(WARN, usage, 0);
				return (1);
			}
			if (!*optarg) {
				strerr_warn1("You cannot have an empty email address", 0);
				strerr_warn2(WARN, usage, 0);
				return (1);
			}
			*option = UPDATE_MEMBER;
			*member = optarg;
			break;
		case 'o':
			if (*option == ADDNEW_GROUP) {
				strerr_warn2(WARN, usage, 0);
				return (1);
			}
			if (!*optarg) {
				strerr_warn1("You cannot have an empty email address", 0);
				strerr_warn2(WARN, usage, 0);
				return (1);
			}
			*option = UPDATE_MEMBER;
			*old_member = optarg;
		case 'h':
			if ((r = scan_int(optarg, &i)) != str_len(optarg))
				i = -1;
			if (!str_diffn(optarg, "DES", 3) || i == DES_HASH)
				strnum[fmt_int(strnum, DES_HASH)] = 0;
			else
			if (!str_diffn(optarg, "MD5", 3) || i == MD5_HASH)
				strnum[fmt_int(strnum, MD5_HASH)] = 0;
			else
			if (!str_diffn(optarg, "SHA-256", 7) || i == SHA256_HASH)
				strnum[fmt_int(strnum, SHA256_HASH)] = 0;
			else
			if (!str_diffn(optarg, "SHA-512", 7) || i == SHA512_HASH)
				strnum[fmt_int(strnum, SHA512_HASH)] = 0;
			else
			if (!str_diffn(optarg, "YESCRYPT", 8) || i == YESCRYPT_HASH)
				strnum[fmt_int(strnum, YESCRYPT_HASH)] = 0;
			else {
				strerr_die5x(100, FATAL, "wrong hash method ", optarg,
						". Supported HASH Methods: DES MD5 SHA-256 SHA-512 YESCRYPT\n", usage);
			}
			if (!env_put2("PASSWORD_HASH", strnum))
				strerr_die1x(111, "out of memory");
			*encrypt_flag = 1;
			break;
		case 'e':
			/*- ignore encrypt flag option if -h option is provided */
			if (*encrypt_flag == -1)
				*encrypt_flag = 0;
			break;
		case 'C':
			if (docram)
				*docram = 1;
			break;
#ifdef HAVE_GSASL
#if GSASL_VERSION_MAJOR == 1 && GSASL_VERSION_MINOR > 8 || GSASL_VERSION_MAJOR > 1
		case 'm':
			if (!scram)
				break;
			if (!str_diffn(optarg, "CRAM", 5))
				*scram = 0;
			else
			if (!str_diffn(optarg, "SCRAM-SHA-1", 11))
				*scram = 1;
			else
			if (!str_diffn(optarg, "SCRAM-SHA-256", 13))
				*scram = 2;
			else
				strerr_die5x(100, WARN, "wrong SCRAM method ", optarg,
						". Supported CRAM Methods: CRAM, SCRAM-SHA1 SCRAM-SHA-256\n", usage);
			break;
		case 'S':
			if (!salt)
				break;
			i = str_chr(optarg, ',');
			if (optarg[i]) {
				strerr_die3x(100, WARN, optarg, ": salt cannot have a comma character");
			}
			*salt = optarg;
			break;
		case 'I':
			if (!iter)
				break;
			scan_int(optarg, iter);
			break;
#endif
#endif
			break;
#ifdef CLUSTERED_SITE
		case 'H':
			*hostid = optarg;
			break;
		case 'M':
			*mdahost = optarg;
			break;
#endif
		default:
			strerr_warn2(WARN, usage, 0);
			return (1);
		}
	}
	if (optind < argc)
		*group = argv[optind++];
	else {
		strerr_warn2(WARN, usage, 0);
		return (1);
	}
	if (optind < argc)
		*passwd = argv[optind++];
	if (*option == UPDATE_MEMBER && (!*member || !*old_member)) {
		strerr_warn2(WARN, usage, 0);
		return (1);
	}
	if (*option == -1) {
		strerr_warn2(WARN, usage, 0);
		return (1);
	}
	if (!*passwd) {
		if (*random)
			*passwd = genpass(*random);
		else
			*passwd = vgetpasswd(*gecos ? *gecos : *group);
	}
	return (0);
}

static void
die_nomem()
{
	strerr_warn1("vgroup: out of memory", 0);
	_exit(111);
}

int
addGroup(char *user, char *domain, char *mdahost, char *gecos,
		char *passwd, char *quota, int encrypt_flag, char *scram)
{
	struct passwd  *pw;
	static stralloc tmpbuf = {0};
	int             i;

	if (!stralloc_copyb(&tmpbuf, "MailGroup ", 10) ||
			!stralloc_cats(&tmpbuf, gecos ? gecos : user) ||
			!stralloc_0(&tmpbuf))
		die_nomem();
	if ((i = iadduser(user, domain, mdahost, passwd, tmpbuf.s, quota, 0, 1, encrypt_flag, scram)) < 0)
		return i;
	if (!(pw = sql_getpw(user, domain))) {
		strerr_warn4(user, "@", domain, ": sql_getpw failed: ", &strerr_sys);
		return 1;
	}
	return 0;
}

int
main(int argc, char **argv)
{
	static stralloc User = {0}, Domain = {0}, alias_line = {0}, old_alias = {0},
					quotaVal = {0}, result = {0};
	char           *group, *gecos, *member, *old_member, *passwd, *hostid,
				   *mdahost, *Quota, *real_domain, *ptr;
	char            strnum[FMT_ULONG];
	mdir_t          q;
	int             i, option, ignore = 0, ret = -1, encrypt_flag, random, docram;
#ifdef HAVE_GSASL
#if GSASL_VERSION_MAJOR == 1 && GSASL_VERSION_MINOR > 8 || GSASL_VERSION_MAJOR > 1
	int             scram, iter;
	char           *b64salt;
#endif
#endif

#ifdef HAVE_GSASL
#if GSASL_VERSION_MAJOR == 1 && GSASL_VERSION_MINOR > 8 || GSASL_VERSION_MAJOR > 1
	i = get_options(argc, argv, &option, &group, &gecos, &member, &old_member,
			&passwd, &hostid, &mdahost, &Quota, &ignore, &encrypt_flag, &random,
			&docram, &scram, &iter, &b64salt);
#else
	i = get_options(argc, argv, &option, &group, &gecos, &member, &old_member,
			&passwd, &hostid, &mdahost, &Quota, &ignore, &encrypt_flag, &random,
			0, 0, 0, 0);
#endif
#else
	i = get_options(argc, argv, &option, &group, &gecos, &member, &old_member,
			&passwd, &hostid, &mdahost, &Quota, &ignore, &encrypt_flag, &random,
			0, 0, 0, 0);
#endif
	if (i)
		return (i);

	parse_email(group, &User, &Domain);
	if (option != ADDNEW_GROUP) {
#ifdef CLUSTERED_SITE
		if (sqlOpen_user(group, 0))
#else
		if (iopen((char *) 0))
#endif
			return (1);
	}
	/* Do this so that users do not get added in a alias domain */
	real_domain = (char *) 0;
	if (Domain.len && !(real_domain = get_real_domain(Domain.s))) {
		strerr_warn2(Domain.s, ": No such domain", 0);
		iclose();
		return (1);
	}
	switch (option)
	{
		case ADDNEW_GROUP:
			if (Quota && *Quota) {
				if (str_diffn(Quota, "NOQUOTA", 8)) {
					if ((q = parse_quota(Quota, 0)) == -1) {
						strerr_warn3("vgroup: parse_quota: ", Quota, ": ", &strerr_sys);
						iclose();
						return (1);
					}
					strnum[i = fmt_ulong(strnum, (unsigned int) q)] = 0;
					if (!stralloc_copyb(&quotaVal, strnum, i))
						die_nomem();
					i = str_chr(Quota, ',');
					if (Quota[i]) {
						if (!stralloc_cats(&quotaVal, Quota + i))
							die_nomem();
					}
					if (!stralloc_0(&quotaVal))
						die_nomem();
					quotaVal.len--;
				}
			} else
				quotaVal.len = 0;
#ifdef CLUSTERED_SITE
			if (!mdahost && hostid) {
				if (!(ptr = sql_getip(hostid))) {
					strerr_warn4("vgroup: failed to obtain mdahost for host ", hostid, " domain ", real_domain, 0);
					iclose();
					return (1);
				} else
					mdahost = ptr;
			}
			/* add the user */
			if (mdahost) {
				iclose();
				if (!(ptr = SqlServer(mdahost, real_domain))) {
					strerr_warn4("vgroup: failed to obtain sqlserver for host ", mdahost, " domain ", real_domain, 0);
					return (1);
				} else
				if (iopen(ptr)) {
					strerr_warn2("vgroup: failed to connect to ", ptr, 0);
					return (1);
				}
				if (verbose)
					strmsg_out5("Connected to MDAhost ", mdahost, " SqlServer ", ptr, "\n");
			}
#endif
#ifdef HAVE_GSASL
#if GSASL_VERSION_MAJOR == 1 && GSASL_VERSION_MINOR > 8 || GSASL_VERSION_MAJOR > 1
			switch (scram)
			{
			case 1: /*- SCRAM-SHA-1 */
				if ((i = gsasl_mkpasswd(verbose, "SCRAM-SHA-1", iter, b64salt, docram, passwd, &result)) != NO_ERR)
					strerr_die2x(111, "gsasl error: ", gsasl_mkpasswd_err(i));
				break;
			case 2: /*- SCRAM-SHA-256 */
				if ((i = gsasl_mkpasswd(verbose, "SCRAM-SHA-256", iter, b64salt, docram, passwd, &result)) != NO_ERR)
					strerr_die2x(111, "gsasl error: ", gsasl_mkpasswd_err(i));
				break;
			/*- more cases will get below when devils come up with a new RFC */
			}
			ptr = scram > 0 ? result.s : 0;
#else
			ptr = 0;
#endif
#else
			ptr = 0;
#endif
			if (!ptr && docram) {
				if (!stralloc_copyb(&result, "{CRAM}", 6) ||
						!stralloc_cats(&result, passwd) ||
						!stralloc_0(&result))
					die_nomem();
				ptr = result.s;
			}
			ret = addGroup(User.s, real_domain, mdahost, gecos,
					passwd, quotaVal.s, encrypt_flag, ptr);
			if (!ret && random) {
				subprintfe(subfdout, "vgroup", "Password is %s\n", passwd);
				flush("vgroup");
			}
			break;
		case INSERT_MEMBER:
			if (*member == '.') {
				strerr_warn2("vgroup: Illegal member ", member, 0);
				iclose();
				return (1);
			}
			if (*member != '&' && *member != '/' && *member != '|') {
				if (!stralloc_copyb(&alias_line, "&", 1) ||
						!stralloc_cats(&alias_line, member) ||
						!stralloc_0(&alias_line))
					die_nomem();
			} else {
				if (!stralloc_copys(&alias_line, member) ||
						!stralloc_0(&alias_line))
					die_nomem();
			}
			alias_line.len--;
			ret = valias_insert(User.s, real_domain, alias_line.s, ignore);
			break;
		case DELETE_MEMBER:
			if (*member == '.') {
				strerr_warn2("vgroup: Illegal member ", member, 0);
				iclose();
				return (1);
			}
			if (*member != '&' && *member != '/' && *member != '|') {
				if (!stralloc_copyb(&alias_line, "&", 1) ||
						!stralloc_cats(&alias_line, member) ||
						!stralloc_0(&alias_line))
					die_nomem();
			} else {
				if (!stralloc_copys(&alias_line, member) ||
						!stralloc_0(&alias_line))
					die_nomem();
			}
			alias_line.len--;
			ret = valias_delete(User.s, real_domain, alias_line.s);
			break;
		case UPDATE_MEMBER:
			if (*member == '.') {
				strerr_warn2("vgroup: Illegal member ", member, 0);
				iclose();
				return (1);
			} else
			if (*old_member == '.') {
				strerr_warn2("vgroup: Illegal member ", old_member, 0);
				iclose();
				return (1);
			}
			if (*member != '&' && *member != '/' && *member != '|') {
				if (!stralloc_copyb(&alias_line, "&", 1) ||
						!stralloc_cats(&alias_line, member) ||
						!stralloc_0(&alias_line))
					die_nomem();
			} else {
				if (!stralloc_copys(&alias_line, member) ||
						!stralloc_0(&alias_line))
					die_nomem();
			}
			alias_line.len--;
			if (*old_member != '&' && *old_member != '/' && *old_member != '|') {
				if (!stralloc_copyb(&old_alias, "&", 1) ||
						!stralloc_cats(&old_alias, member) ||
						!stralloc_0(&old_alias))
					die_nomem();
			} else {
				if (!stralloc_copys(&old_alias, member) ||
						!stralloc_0(&old_alias))
					die_nomem();
			}
			old_alias.len--;
			ret = valias_update(User.s, real_domain, old_alias.s, alias_line.s);
			break;
	}
	iclose();
	return (ret);
}
#else
#include <strerr.h>
int
main()
{
	strerr_warn1("IndiMail not configured with --enable-valias=y", 0);
	return (0);
}
#endif

/*
 * $Log: vgroup.c,v $
 * Revision 1.8  2023-07-17 12:28:44+05:30  Cprogrammer
 * added YESCRYPT hash
 *
 * Revision 1.7  2023-07-15 00:26:13+05:30  Cprogrammer
 * Set password for CRAM authentication with {CRAM} prefix.
 *
 * Revision 1.6  2023-01-22 10:40:03+05:30  Cprogrammer
 * replaced qprintf with subprintf
 *
 * Revision 1.5  2022-11-02 20:15:00+05:30  Cprogrammer
 * added feature to add scram password during user addition
 *
 * Revision 1.4  2022-08-05 22:44:59+05:30  Cprogrammer
 * removed apop argument to iadduser()
 * added encrypt flag to iadduser()
 *
 * Revision 1.3  2019-06-07 15:51:55+05:30  mbhangui
 * use sgetopt library for getopt()
 *
 * Revision 1.2  2019-04-22 23:19:29+05:30  Cprogrammer
 * added missing strerr.h
 *
 * Revision 1.1  2019-04-18 08:34:02+05:30  Cprogrammer
 * Initial revision
 *
 */
